hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java

```diff
@@ -1033,6 +1035,23 @@ else if ( CollectionPart.Nature.fromNameExact( parentNavigablePath.getLocalName(
 		}
 	}
 
+	private NavigablePath getReferencedNavigablePath(
+			DomainResultCreationState creationState,
+			NavigablePath parentNavigablePath) {
+		NavigablePath referencedNavigablePath = parentNavigablePath.getParent();
+		MappingType partMappingType = creationState.resolveModelPart( referencedNavigablePath ).getPartMappingType();
+
+		while ( !( partMappingType instanceof EntityPersister )
+				|| ( partMappingType != entityMappingType
+				&& !entityMappingType.isSubclassEntityName( partMappingType.getMappedJavaType().getJavaType().getTypeName() )
+				&& !( (EntityPersister) partMappingType ).isSubclassEntityName( entityMappingType.getEntityName() ) ) ) {
```

Comment:
Would be awesome if you could add a little comment with an example to make it easier to understand what is going on here. I'm actually a bit surprised that the model shown in the test triggers an issue. I would assume we have plenty of similar models.

Reply:
I'll add a comment